File: pkg/shmevent/logpermit.go
package shmevent

import "fmt"

// LogPermitKey builds the pkg/store key for a KindLogPermit record:
// SystemKeyPrefix, KindLogPermit, status, then a 2-byte big-endian length
// prefix for logKind, then logKind, then peerID verbatim. Unlike
// SystemKey, two variable-length fields follow the fixed header here
// (logKind and peerID), so only the last of them -- peerID -- can go
// unprefixed; logKind needs its own length prefix so the two can't blur
// into each other (the same fix pkg/logrecord.BuildKey applies to its own
// kind/unitID fields, for the identical reason).
func LogPermitKey(status byte, logKind string, peerID []byte) ([]byte, error) {
	if len(logKind) > 0xFFFF {
		return nil, fmt.Errorf("shmevent: log permit logKind too long: %d bytes", len(logKind))
	}
	key := make([]byte, 3+2+len(logKind)+len(peerID))
	key[0] = SystemKeyPrefix
	key[1] = KindLogPermit
	key[2] = status
	key[3] = byte(len(logKind) >> 8)
	key[4] = byte(len(logKind))
	off := 5
	off += copy(key[off:], logKind)
	copy(key[off:], peerID)
	return key, nil
}

// EncodeLogPermitRequestPayload packs logKind, peerID, and metadata into a
// single EventLogPermitRequest Msg.Value: a 2-byte big-endian length
// prefix for logKind, then logKind, then a 2-byte big-endian length
// prefix for peerID, then peerID, then metadata verbatim (the rest of the
// buffer, needing no length prefix of its own) -- mirrors
// EncodePermitRequestPayload, except logKind is a string needing its own
// prefix (EncodePermitRequestPayload's kind is a single fixed byte) in
// place of that single kind byte.
func EncodeLogPermitRequestPayload(logKind string, peerID, metadata []byte) ([]byte, error) {
	if len(logKind) > 0xFFFF {
		return nil, fmt.Errorf("shmevent: log permit request logKind too long: %d bytes", len(logKind))
	}
	if len(peerID) > 0xFFFF {
		return nil, fmt.Errorf("shmevent: log permit request peerID too long: %d bytes", len(peerID))
	}
	buf := make([]byte, 2+len(logKind)+2+len(peerID)+len(metadata))
	buf[0] = byte(len(logKind) >> 8)
	buf[1] = byte(len(logKind))
	off := 2
	off += copy(buf[off:], logKind)
	buf[off] = byte(len(peerID) >> 8)
	buf[off+1] = byte(len(peerID))
	off += 2
	off += copy(buf[off:], peerID)
	copy(buf[off:], metadata)
	return buf, nil
}

// DecodeLogPermitRequestPayload is the inverse of
// EncodeLogPermitRequestPayload.
func DecodeLogPermitRequestPayload(payload []byte) (logKind string, peerID, metadata []byte, err error) {
	if len(payload) < 2 {
		return "", nil, nil, fmt.Errorf("shmevent: log permit request payload too short: %d bytes", len(payload))
	}
	kindLen := int(payload[0])<<8 | int(payload[1])
	off := 2
	if off+kindLen > len(payload) {
		return "", nil, nil, fmt.Errorf("shmevent: log permit request logKind length %d exceeds payload size %d", kindLen, len(payload))
	}
	logKind = string(payload[off : off+kindLen])
	off += kindLen
	if off+2 > len(payload) {
		return "", nil, nil, fmt.Errorf("shmevent: log permit request payload too short for peerID length")
	}
	idLen := int(payload[off])<<8 | int(payload[off+1])
	off += 2
	if off+idLen > len(payload) {
		return "", nil, nil, fmt.Errorf("shmevent: log permit request peerID length %d exceeds payload size %d", idLen, len(payload))
	}
	return logKind, payload[off : off+idLen], payload[off+idLen:], nil
}

// EncodeLogPermitConfirmPayload packs logKind and peerID (the rest of the
// buffer) into a single EventLogPermitConfirm/EventLogPermitRevoke
// Msg.Value -- no metadata field, mirroring EncodePermitConfirmPayload's
// reasoning (the daemon reads the pending record's own value back out of
// the store rather than trusting the caller to resend it), with logKind's
// own length prefix taking the place of EncodePermitConfirmPayload's
// single kind byte.
func EncodeLogPermitConfirmPayload(logKind string, peerID []byte) ([]byte, error) {
	if len(logKind) > 0xFFFF {
		return nil, fmt.Errorf("shmevent: log permit confirm logKind too long: %d bytes", len(logKind))
	}
	buf := make([]byte, 2+len(logKind)+len(peerID))
	buf[0] = byte(len(logKind) >> 8)
	buf[1] = byte(len(logKind))
	off := 2
	off += copy(buf[off:], logKind)
	copy(buf[off:], peerID)
	return buf, nil
}

// DecodeLogPermitConfirmPayload is the inverse of
// EncodeLogPermitConfirmPayload.
func DecodeLogPermitConfirmPayload(payload []byte) (logKind string, peerID []byte, err error) {
	if len(payload) < 2 {
		return "", nil, fmt.Errorf("shmevent: log permit confirm payload too short: %d bytes", len(payload))
	}
	kindLen := int(payload[0])<<8 | int(payload[1])
	off := 2
	if off+kindLen > len(payload) {
		return "", nil, fmt.Errorf("shmevent: log permit confirm logKind length %d exceeds payload size %d", kindLen, len(payload))
	}
	return string(payload[off : off+kindLen]), payload[off+kindLen:], nil
}

File: web-app/src/shmevent/logpermit.rs
//! Rust port of `pkg/shmevent/logpermit.go`: `KindLogPermit` record
//! key-building and payload framing -- a permit scoped by an arbitrary
//! `logKind` string in addition to `peerID`, unlike [`super::system`]'s
//! fixed-kind-byte permits.

use super::system::{KIND_LOG_PERMIT, SYSTEM_KEY_PREFIX};
use super::Error;

/// Builds the store key for a `KindLogPermit` record: `SYSTEM_KEY_PREFIX`,
/// `KIND_LOG_PERMIT`, `status`, then a 2-byte big-endian length prefix for
/// `log_kind`, then `log_kind`, then `peer_id` verbatim. Matches
/// `pkg/shmevent.LogPermitKey`.
pub fn log_permit_key(status: u8, log_kind: &str, peer_id: &[u8]) -> Result<Vec<u8>, Error> {
    let log_kind_bytes = log_kind.as_bytes();
    if log_kind_bytes.len() > 0xFFFF {
        return Err(Error(format!(
            "log permit logKind too long: {} bytes",
            log_kind_bytes.len()
        )));
    }
    let mut key = Vec::with_capacity(3 + 2 + log_kind_bytes.len() + peer_id.len());
    key.push(SYSTEM_KEY_PREFIX);
    key.push(KIND_LOG_PERMIT);
    key.push(status);
    key.push((log_kind_bytes.len() >> 8) as u8);
    key.push(log_kind_bytes.len() as u8);
    key.extend_from_slice(log_kind_bytes);
    key.extend_from_slice(peer_id);
    Ok(key)
}

/// Packs `log_kind`, `peer_id`, and `metadata` into a single
/// `EVENT_LOG_PERMIT_REQUEST` `Msg.value`: a 2-byte big-endian length
/// prefix for `log_kind`, then `log_kind`, then a 2-byte big-endian length
/// prefix for `peer_id`, then `peer_id`, then `metadata` verbatim (the
/// rest of the buffer). Matches
/// `pkg/shmevent.EncodeLogPermitRequestPayload`.
pub fn encode_log_permit_request_payload(
    log_kind: &str,
    peer_id: &[u8],
    metadata: &[u8],
) -> Result<Vec<u8>, Error> {
    let log_kind_bytes = log_kind.as_bytes();
    if log_kind_bytes.len() > 0xFFFF {
        return Err(Error(format!(
            "log permit request logKind too long: {} bytes",
            log_kind_bytes.len()
        )));
    }
    if peer_id.len() > 0xFFFF {
        return Err(Error(format!(
            "log permit request peerID too long: {} bytes",
            peer_id.len()
        )));
    }
    let mut buf =
        Vec::with_capacity(2 + log_kind_bytes.len() + 2 + peer_id.len() + metadata.len());
    buf.push((log_kind_bytes.len() >> 8) as u8);
    buf.push(log_kind_bytes.len() as u8);
    buf.extend_from_slice(log_kind_bytes);
    buf.push((peer_id.len() >> 8) as u8);
    buf.push(peer_id.len() as u8);
    buf.extend_from_slice(peer_id);
    buf.extend_from_slice(metadata);
    Ok(buf)
}

/// Inverse of [`encode_log_permit_request_payload`].
pub fn decode_log_permit_request_payload(
    payload: &[u8],
) -> Result<(String, &[u8], &[u8]), Error> {
    if payload.len() < 2 {
        return Err(Error(format!(
            "log permit request payload too short: {} bytes",
            payload.len()
        )));
    }
    let kind_len = ((payload[0] as usize) << 8) | payload[1] as usize;
    let mut off = 2;
    if off + kind_len > payload.len() {
        return Err(Error(format!(
            "log permit request logKind length {kind_len} exceeds payload size {}",
            payload.len()
        )));
    }
    let log_kind = String::from_utf8_lossy(&payload[off..off + kind_len]).to_string();
    off += kind_len;
    if off + 2 > payload.len() {
        return Err(Error(
            "log permit request payload too short for peerID length".to_string(),
        ));
    }
    let id_len = ((payload[off] as usize) << 8) | payload[off + 1] as usize;
    off += 2;
    if off + id_len > payload.len() {
        return Err(Error(format!(
            "log permit request peerID length {id_len} exceeds payload size {}",
            payload.len()
        )));
    }
    Ok((log_kind, &payload[off..off + id_len], &payload[off + id_len..]))
}

/// Packs `log_kind` and `peer_id` (the rest of the buffer) into a single
/// `EVENT_LOG_PERMIT_CONFIRM`/`EVENT_LOG_PERMIT_REVOKE` `Msg.value` -- no
/// metadata field, mirroring `encode_permit_confirm_payload`'s reasoning.
/// Matches `pkg/shmevent.EncodeLogPermitConfirmPayload`.
pub fn encode_log_permit_confirm_payload(log_kind: &str, peer_id: &[u8]) -> Result<Vec<u8>, Error> {
    let log_kind_bytes = log_kind.as_bytes();
    if log_kind_bytes.len() > 0xFFFF {
        return Err(Error(format!(
            "log permit confirm logKind too long: {} bytes",
            log_kind_bytes.len()
        )));
    }
    let mut buf = Vec::with_capacity(2 + log_kind_bytes.len() + peer_id.len());
    buf.push((log_kind_bytes.len() >> 8) as u8);
    buf.push(log_kind_bytes.len() as u8);
    buf.extend_from_slice(log_kind_bytes);
    buf.extend_from_slice(peer_id);
    Ok(buf)
}

/// Inverse of [`encode_log_permit_confirm_payload`].
pub fn decode_log_permit_confirm_payload(payload: &[u8]) -> Result<(String, &[u8]), Error> {
    if payload.len() < 2 {
        return Err(Error(format!(
            "log permit confirm payload too short: {} bytes",
            payload.len()
        )));
    }
    let kind_len = ((payload[0] as usize) << 8) | payload[1] as usize;
    let off = 2;
    if off + kind_len > payload.len() {
        return Err(Error(format!(
            "log permit confirm logKind length {kind_len} exceeds payload size {}",
            payload.len()
        )));
    }
    let log_kind = String::from_utf8_lossy(&payload[off..off + kind_len]).to_string();
    Ok((log_kind, &payload[off + kind_len..]))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn log_permit_key_layout() {
        let key = log_permit_key(0x01, "cmdlog", b"peer-abc").unwrap();
        assert_eq!(key[0], SYSTEM_KEY_PREFIX);
        assert_eq!(key[1], KIND_LOG_PERMIT);
        assert_eq!(key[2], 0x01);
        let kind_len = ((key[3] as usize) << 8) | key[4] as usize;
        assert_eq!(kind_len, 6);
        assert_eq!(&key[5..5 + kind_len], b"cmdlog");
        assert_eq!(&key[5 + kind_len..], b"peer-abc");
    }

    #[test]
    fn log_permit_request_payload_round_trip() {
        let encoded =
            encode_log_permit_request_payload("cmdlog", b"peer-123", b"meta").unwrap();
        let (log_kind, peer_id, metadata) = decode_log_permit_request_payload(&encoded).unwrap();
        assert_eq!(log_kind, "cmdlog");
        assert_eq!(peer_id, b"peer-123");
        assert_eq!(metadata, b"meta");
    }

    #[test]
    fn log_permit_confirm_payload_round_trip() {
        let encoded = encode_log_permit_confirm_payload("cmdlog", b"peer-456").unwrap();
        let (log_kind, peer_id) = decode_log_permit_confirm_payload(&encoded).unwrap();
        assert_eq!(log_kind, "cmdlog");
        assert_eq!(peer_id, b"peer-456");
    }

    #[test]
    fn log_permit_confirm_payload_too_short_rejected() {
        assert!(decode_log_permit_confirm_payload(&[0u8]).is_err());
    }
}
